Implement capture sharpening inside demosaic module

jenshannoschwalm opened the pull request:
This has been discussed many times here and on pixls.us so here we go. As we have D&S giving good results i was somewhat reluctant to work on this and do a pr but as CS works so good on RT and after a lot of testing (and using this on my personal images for a while with 3 auto applied presets for my fixed-lens camera - one for high-iso with CS turned off and two presets for wide open and "normal") i am pretty confident it's worth it and would be a significant improvement.
(It took quite a while to spot a very bad bug in GPU code leading to system instability and crashes so unfortunately not for 5.2)

Please test in depth and let's decide if this is for master. It's a lot of processing - so not a superfast module.

_________________________________________________
Capture sharpening has been implemented to work inside the demosaic module so it's raw only.
Credits to: Ingo Weyrich (heckflosse67@gmx.de), he implemented the original algorithm for rawtherapee, this implementation is based on his work, especially the convolution kernels.
CPU and OpenCL code paths are both available.
Demosaic module gets more parameters so there is a version bump, one float has been reserved.

**The "mini manual"**

Capture sharpening (CS) tries to recover details lost due to in-camera blurring, which can be caused by diffraction, the anti-aliasing filter or other sources of gaussian-type blur. Prerequisites are
- good white balance parameters (same requirement as for highlights reconstruction or demosaic)
- no chromatic aberration, you might want to add the "raw chromatic aberration" module
- low noise as noise will be amplified by CS

**Controls**
1. **capture sharpen** switches CS on if above zero and defines the strength of overall effect. CS works in an iterative process, this defines the number of iterations. As downscaling after demosaicing reduces the visibility of CS we reduce iterations in such cases for performance.
2. **radius** defines the basic convolution gaussian sigma. This should not be set by "creative means" but to the blurring radius of the optical system and sensor. Values up to 1 are (almost always) safe, if set too high halos will be introduced.
3. **contrast threshold** As sensor noise will be amplified by CS we take some care about this by a per pixel variance analysis and use a logistic function with this threshold to avoid CS in noisy areas.
4. **corner boost** Increase the radius in image corners. We assume a circle of 1/2 of image size to be safe (only use main radius), locations outside this center circle get an increased convolution radius.

Planned after deciding about integration into master:
1. For sure: Auto-calculate the radius
2. Possibly: some halo suppression and luma denoising

da-phil commented:
Nice effort, I like it :+1: 
However for some images I noticed a color shift the more iterations I choose for CS. Can you also reproduce it? Try very colorful images and maybe bump up the saturation a little bit in order to make the difference more visible.
In my case I also overlayed the non-CS with the CS image and used the layer mode "difference" to also see the color cast quantitatively.

jenshannoschwalm commented:
> However for some images I noticed a color shift the more iterations I choose for CS. Can you also reproduce it?

I couldn't yet, could you somehow share such raw files? (And possibly the xmp too? It might be related to white balance coeffs. A pre-requisite would be a correct WB)

As CS is an iterative process any error would be amplified by more iterations.


> In my case I also overlayed ...

I don't understand your "reproducer procedure" here.

jenshannoschwalm commented:
I can do so of course.

da-phil commented:
> > In my case I also overlayed ...
> 
> I don't understand your "reproducer procedure" here.

What I meant was exporting an image without CS and one with CS turned on with at least 15 iterations, then compare the images overlayed in separate layers in gimp, using the layer mode "difference".
For some images I also noticed that the image with CS turned on appear slightly brighter.

I'll try to find images which I can share for you to reproduce.

jenshannoschwalm commented:
> What I meant was exporting an image without CS and one with CS turned on with at least 15 iterations, then compare the images overlayed in separate layers in gimp, using the layer mode "difference".

You are aware of the `--dump-pipe module` option?

jenshannoschwalm commented:
Squashed and force-pushed an updated version
1. convolution kernels support sigma-steps of 0.01 instead of 0.02
2. slightly improved CPU convolution performance
3. the per-pixel variance anylysis is done in a circle with a diameter of 5 instead of 3 for improved stability in noisy regions and at sharp luminance transitions.

jenshannoschwalm commented:
> > > What I meant was exporting an image without CS and one with CS turned on with at least 15 iterations, then compare the images overlayed in separate layers in gimp, using the layer mode "difference".
> > 
> > 
> > You are aware of the `--dump-pipe module` option?
> 
> Nope, never used it, what is it good for?

It writes pfm files to tmp directory while processing the specified module in the pipe, it's meant for debugging exactly such issues :-)

`darktable --dump-pipe demosaic` would be your friend (use HQ processing mode)

jenshannoschwalm commented:
Latest version includes a variance lut for performance plus some readability maintenance. (Some NaN issues spotted while testing this are in a bugfix PR for 5.2)

jenshannoschwalm commented:
Latest version got some subtle perf gain, more important is improved stability for high-noise underexposed images. Right now i think it's good as is and i would appreciate testing. Please bear in mind, i have squashed and force-pushed the update to keep number of commits low.

I am still not convinced about the UI.
1. Is it really better to make the unused sliders insensitive instead of just hiding?
2. Any other idea?

jenshannoschwalm commented:
Normally with a strength of 10 would be fine. Evaluate at 100% zoom.

jenshannoschwalm commented:
1. use a different default contrast threshold after a lot of testing - definitely better for medium ISO images while still perfect for low ISO
2. Second commit while being here cleans up some demosaicer code.

jenshannoschwalm commented:
Minor code change to make params check happy.

jenshannoschwalm commented:
Some subtle const improvements have been force-pushed

jenshannoschwalm commented:
To be reviewed: Instead of using a slider defining the iterations use a simple menu.
Also hide capture sharpen parameters if not applicable or unused as we do for other demosaicer internals.

This would be **my** personal preference.

da-phil commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+  global float *d = in + i;
+
+  float val = 0.0f;
+  if(col >= 4 && row >= 4 && col < w1 - 4 && row < height - 4)
```
Prefer using braces to make operation order more readable and less error-prone

jenshannoschwalm replied:
Generally yes

da-phil commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+{
+  const int col = get_global_id(0);
+  const int row = get_global_id(1);
+  if(col >= w1 || row >= height) return;
```
Do we have a global formatting style? I see inconsistencies here and there, like keeping the if branch statement in the the same line as the condition and in line 106 it's in ithe next line.

jenshannoschwalm replied:
In general i think we prefer "next line and indent", for such "shortcuts" we also accept the dense style.

da-phil replied:
I mean we even have a `.clang-format` file lurking around in the repo root folder, I always wondered why we do not just stick to consistent formatting through clang-format and include a CI job to enforce it.

da-phil commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+  if(blendmask[k] > CAPTURE_BLEND_EPS)
+  {
+    const float mixer = clamp(blendmask[k], 0.0f, 1.0f);
+    const float lumold = fmax(luminance[k], 0.000001f);
```
Also those magic constants are interesting ;)

jenshannoschwalm replied:
1. `CAPTURE_BLEND_EPS` is emperical and just avoids iterations per pixel if the gaussian blurred mask is "pretty small"
2. Again there is no math background for `0.000001f` just something very small and positive.

da-phil replied:
`CAPTURE_BLEND_EPS` is great and the meaning obvious, I just meant the numeric literals.
Just a dump question, can you not just use `CAPTURE_BLEND_EPS` instead of the literal?

weltyj commented:
@jenshannoschwalm  This is really, really good.   I ran some tests against the pixel deblur module I was working on and this is clearly an order of magnitude better than pixeldeblur.   Amazing how it does not exaggerate noise.   Halos and ringing can be a problem if you push it too far (I did say I was *testing* ;-) ).   I'll  have a closer look as time goes by (busy week coming up).

One thing I would choose to change is the "corner radius" implementation.   Every lens is going to have different focus/diffraction as a function of distance from the image center.

jenshannoschwalm commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+#define CAPTURE_YMIN 0.001f
+#define CAPTURE_THRESHPOWER 0.15f
+
+static inline float sqrf(float a)
```
ok, renamed to fsquare() and - as used elsewhere - moved to common.h

jenshannoschwalm commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+
+#include "common.h"
+
+#define CAPTURE_KERNEL_ALIGN 32
```
To me, instead of doing `const int` in several functions it seems better/safer to use #defines here so we could change without hassle.

jenshannoschwalm commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+  if(col >= w || row >= height) return;
+
+  float4 rgb = read_imagef(dev_out, samplerA, (int2)(col, row));
+  const float4 flum = (float4)( 0.212671f, 0.715160f, 0.072169f, 0.0f );
```
We could :-) At least in cpu code we had a comment what this is about.

da-phil replied:
Can we share global constants across the CPU & OpenCL code?

jenshannoschwalm commented:
force-pushed after review-suggested mods.

jenshannoschwalm commented:
> Halos and ringing can be a problem if you push it too far

Right. No idea yet though how/if we could control this further.

> One thing I would choose to change is the "corner radius" implementation. Every lens is going to have different focus/diffraction as a function of distance from the image center.

True. The good point in this implementation is, we can easily adopt for other and improved algorithms as we use a per-pixel map defining the local gaussian sigma, see `_cs_precalc_gauss_idx`. I would love to see improvements happening after this hopefully get's merged :-)

I could add a "visualizer" to that parameter ...

jenshannoschwalm commented:
Added a visualizing button to corner boost ...

da-phil commented on data/kernels/capture.cl:
```diff
@@ -0,0 +1,287 @@
+                             const int w1,
+                             const int height)
+{
+  const int col = get_global_id(0);
```
Yes, I meant something like 
```suggestion
  // as per your preference:
  #define IMG_COLUMNS 0
  // ... or...
  const int IMG_COLUMNS = 1;
  // ...
  const int col = get_global_id(IMG_COLUMNS);
```

da-phil commented:
> To be reviewed: Instead of using a slider defining the iterations use a simple menu. Also hide capture sharpen parameters if not applicable or unused as we do for other demosaicer internals.
> 
> This would be **my** personal preference.

I'm fine with hiding the capture sharpen parameters as it is currently done. Having a drop-down menu with presets including "disable" is an elegant way to save one UI element, it feels more natural than the previous slider ;)

However I'm a little bit torn with having only "presets" instead of actually defining the amount of iterations, as in RawTherapee. As I'd consider DT users (including myself) as rather tech savy, I'd prefer to fine-tune iterations, and maybe slightly go above the extreme iteration value in some cases.
So I wouldn't mind the extra UI element, and make the drop-down menu only switch CS on / off and have the iterations slide back.

> Added a visualizing button to corner boost ...
Thanks, looks good :+1:

weltyj commented:
A speedup to _modify_blend (much more efficient way to calculate standard deviation).   I'll just put the code for the function here.   This is from src/iop/demosaicing/capture.c 

static void _modify_blend(float *blend,
                          float *Yold,
                          float *luminance,
                          const float threshold,
                          const int width,
                          const int height)
{
  DT_OMP_FOR()
  for(int irow = 0; irow < height; irow++)
  {
    const int row = CLAMP(irow, 2, height-3);
    for(int icol = 0; icol < width; icol++)
    {
      const int col = CLAMP(icol, 2, width-3);
      const size_t k = (size_t)irow * width + icol;
      float sum = 0.0f;
      float sum_sq = 0.0f;
      int n=0 ;

      for(int y = row-1; y < row+2; y++)
      {
        for(int x = col-2; x < col+3; x++) {
          sum += Yold[(size_t)y*width + x];
          sum_sq += Yold[(size_t)y*width + x]*Yold[(size_t)y*width + x] ;
	  n++ ;
	}
      }

      for(int x = col-1; x < col+2; x++)
      {
          sum += Yold[(size_t)(row-2)*width + x];
          sum_sq += Yold[(size_t)(row-2)*width + x]*Yold[(size_t)(row-2)*width + x] ;
	  n++ ;
          sum += Yold[(size_t)(row+2)*width + x];
          sum_sq += Yold[(size_t)(row+2)*width + x]*Yold[(size_t)(row+2)*width + x] ;
	  n++ ;
      }

      const float sum_of_squares = sum_sq - sum*sum/(float)n ;
      const float variance = sum_of_squares / (float)n ;
      const float std_deviation = (variance) ;
      float sv = powf(MAX(0.0f, 5.0f * std_deviation - threshold), CAPTURE_THRESHPOWER);
      blend[k] *= CLIP(sv);
      luminance[k] = Yold[k];
    }
  }
}

weltyj commented:
Two things I've learned since yesterday using my test images from pixeldeblur:

1) Code as is fails badly on underexposed images.   I have a better result using coefficient of variation instead of standard deviation in _modify_blend(), I'm still playing with how to scale it and apply the threshold.

2) The "ringing" effect is caused because the value of the blend[]  mask drops too quickly from areas of high contrast to areas of low contrast.   It may require a blur on the blend[] mask after _modify_blend().

A random thought:   What would happen if the same pixel radius was always used, but in the iteration loop, the blend mask was changed on a per pixel basis -- so some pixel locations would stop having sharpening effects applied sooner?

Out on a 23k hike tomorrow at Mt. Rainier.   Wildflowers are blooming!

jenshannoschwalm commented:
Again fixed, squashed and force-pushed an updated version, also updated the initial mini manual after comments and review.

@weltyj used the faster variance code. Would you be able to share some images from your test image collection? 

> 2. The "ringing" effect is caused because the value of the blend[]  mask drops too quickly from areas of high contrast to areas of low contrast.   It may require a blur on the blend[] mask after _modify_blend().

That might be helping indeed.

> 1. Code as is fails badly on underexposed images.   I have a better result using coefficient of variation instead of standard deviation in _modify_blend(), I'm still playing with how to scale it and apply the threshold.

You mean too much added chroma noise?

@da-phil `CAPTURE_YMIN` can and should be used instead of the 0.00001f magic. Also fixed subtle differences between CPU and OpenCL code paths.

jenshannoschwalm commented:
I tested with sigma 4.0 - 6.0, all are better :-) pr following

jenshannoschwalm commented:
@weltyj Hope you had a fine hike, haven't been there because of restricted time and was fascinated too much in olympic :-) You were of course right about the blending mask. Reimplemented a logistic function on std_deviation, it took some time to find good 'magics' as i was not sure what values we might have for variance/std_deviation. Also made sure we definitely reach zero for chosen-high-variance pixels (for performance as we can avoid the convolution). Would you try - for me working good now?

da-phil commented:
Tested the latest version and have the following observations:
* When I click on the icon to calculate the radius I get deviating values as pop-up messages, see screen capture
https://github.com/user-attachments/assets/dada7ecb-968d-4f03-92bd-e406d0719fce
* Also the contrast threshold seems to be behave differently for HDR DNG files (generated by darktable HDR function), with the default value of 35% barely anything except the sharpest edges are selected
<img width="2413" height="1537" alt="image" src="https://github.com/user-attachments/assets/e1220438-84a3-4d21-b6c3-84774674fee6" />
* I'm not sure I understand how the "sharp center" slider works, it seems to affect the corner boost mask if corner boost was increase significantly.

weltyj commented:
@da-phil I hope the problem with your test image having only the sharpest edges selected will be fixed as the algorithm moves to coefficient of variation instead of standard deviation for detection of high contrast.   I saw exactly the same thing (only worse!) when I looked at one of my underexposed images.

MStraeten commented:
there seems to be an issue with the opencl kernel on macOS
darktable -d opencl -d memory:
   ```
    40.2630 [memory] before pixelpipe process
            max address space (vmpeak):         unknown
            cur address space (vmsize):    415077216 kB
            max used memory   (vmhwm ):         unknown
            cur used memory   (vmrss ):      1743808 kB
    40.2657 pipe cache check              [full]                                        64 lines (important=5, used=8). Freed: invalid 193MB used 0MB. Using 430MB, limit=1024MB
    40.2657 [memory] before pixelpipe process
            max address space (vmpeak):         unknown
            cur address space (vmsize):    415077024 kB
            max used memory   (vmhwm ):         unknown
            cur used memory   (vmrss ):      1752976 kB
    40.2658 [opencl memory] device 'Apple Apple M1 Max' id=0: 92.4MB in use, 10789.8MB available GPU mem of 16384.0MB
    ...
    40.2794 [opencl memory] device 'Apple Apple M1 Max' id=0: 510.4MB in use, 10789.8MB available GPU mem of 16384.0MB
    40.3228 cache report                  [preview]                                     12 lines (important=3, used=10, invalid=0). Using 163MB, limit=0MB. Hits/run=0.67. Hits/test=0.083
    40.3228 [memory] after pixelpipe process
            max address space (vmpeak):         unknown
            cur address space (vmsize):    416691952 kB
            max used memory   (vmhwm ):         unknown
            cur used memory   (vmrss ):      1762320 kB
    40.3898 [opencl memory] device 'Apple Apple M1 Max' id=0: 602.8MB in use, 10789.8MB available GPU mem of 16384.0MB
    ...
    40.3902 [dt_opencl_enqueue_kernel_2d_args_internal] kernel `(null)' (262) on device 'Apple Apple M1 Max' id=0: CL_INVALID_KERNEL
    40.3902 capture sharpen failed    CL0 [full]           demosaic                900  Error: CL_INVALID_KERNEL
    40.3902 [opencl memory] device 'Apple Apple M1 Max' id=0: 1157.3MB in use, 10789.8MB available GPU mem of 16384.0MB
    ...
    40.3903 [opencl memory] device 'Apple Apple M1 Max' id=0: 140.7MB in use, 10789.8MB available GPU mem of 16384.0MB
    40.3903 Error: process            CL0 [full]           demosaic                900       (0/0)  6022x4023 sc=1.000 -->       (0/0)  2176x1454 sc=0.361; device=0 (appleapplem1max), CL_INVALID_KERNEL
    40.3903 pipe aborts               CL0 [full]           demosaic                900       (0/0)  6022x4023 sc=1.000 -->       (0/0)  2176x1454 sc=0.361; couldn't run module on GPU, falling back to CPU
    40.3903 [opencl memory] device 'Apple Apple M1 Max' id=0: 92.4MB in use, 10789.8MB available GPU mem of 16384.0MB
    ...
    41.3464 [opencl memory] device 'Apple Apple M1 Max' id=0: 0.0MB in use, 10789.8MB available GPU mem of 16384.0MB
    41.3483 cache report                  [full]                                        64 lines (important=7, used=15, invalid=3). Using 768MB, limit=1024MB. Hits/run=0.67. Hits/test=0.074
    41.3484 [memory] after pixelpipe process
            max address space (vmpeak):         unknown
            cur address space (vmsize):    415472496 kB
            max used memory   (vmhwm ):         unknown
            cur used memory   (vmrss ):      2085712 kB
    41.3700 pipe cache check              [preview]                                     12 lines (important=3, used=10). Freed: invalid 0MB used 0MB. Using 163MB, limit=0MB```

jenshannoschwalm commented:
> seems to conflict with color smoothing:

Will have a look, not yet observed here



> there seems to be an issue with the opencl kernel on macOS

Could you remove all kernels and enforce recompilation - i know there have been issues on Macs - with `-d opencl -d verbose` to see compilation logs ?

weltyj commented:
>       const float modified_coef_variation = std_deviation/sqrtf(mean) ;
>       float y = 1./(1.+expf(5.0f-1000.0f*(modified_coef_variation-threshold/0.06))) ;
> 
>       // this bit of logic will remap the values from (ytrunc to 1.0) to (0.0 to 1.0)
>       float ytrunc = 1./(1.+exp(5.0)) ;  // ~ 0.0067 -- the untruncated logistic evaluated at 0.0
>       float sv = (y-ytrunc)/(1-ytrunc) ;
>       blend[k] *= CLIP(sv);

@jenshannoschwalm I just tried that, looks pretty good.   Give it try see what you think.

jenshannoschwalm commented:
A few more comments:
1. We still want to do some normalizing on input signal range. That will definitely further improve stability of variance analysis for severely underexposed images or for HDR/DNG images as reported by @da-phil . Currently i think we should **not** try to "fix further" by magic in capture code. There is pending pixelpipe/demosaic stuff here that should be integrated before.
2. To me raw denoising module seems to be more "sort of a problem" than a "problem solver"
3. As mentioned above i will later do some PR's related to demosaicing, one part is a far better dual demosaicing that will also improve CS for very noisy images so let's not tackle this here.

BTW - big thanks to all your feedback/review/discussions.

da-phil commented:
> 1. We still want to do some normalizing on input signal range. That will definitely further improve stability of variance analysis for severely underexposed images or for HDR/DNG images as reported by @da-phil . Currently i think we should **not** try to "fix further" by magic in capture code. There is pending pixelpipe/demosaic stuff here that should be integrated before.

This means that you are not going to use the coefficient of variation instead of the std var as a means to normalize variation against the mean of the signal?

weltyj commented:
> About the HDR image from @da-phil, technically it also has very low signals later corrected in exposure module by ~3EV. At least with the new variance analysis we can "make it fit"

@jenshannoschwalm I want to make sure I understand.     Are you saying that capture.c as it is (in your github fork) correctly identifies the edges to be sharpened, even though it is about -3EV from a normal exposure?

jenshannoschwalm commented:
@weltyj and @da-phil i am very pleased with current state of code and results. There are some things i'd like to discuss before "letting this into the wild"
1. Not sure if we can avoid some processing for low-blend values. We "often" iterate at those locations but nothing significant can be seen. That currently would be `CAPTURE_BLEND_EPS`.
2. About the "low-exposure" problem. Many out-of-camera raw files have tags `Exif.Photo.ExposureBiasValue` or `Exif.Image.ExposureBiasValue` and those are read in `exif.cc` and stored in `img->exif_exposure_bias`. It's not as good as an analysed maximum signal but i am sure we could use that to correct out variance-sigmoid-analysis so low exposure images behave better. Unfortunately that leaves out images as the one from @da-phil

jenshannoschwalm commented:
Btw you seem to have misunderstood the eps level? That is only used as a threshold to avoid the convolution per pixel instead of iteration with very small changes.

weltyj commented:
@jenshannoschwalm Think about this:  2 loops in modify_blend.  The first loop computes CV for each pixel, and collects meta data about the mean, min,max and standard deviation **of the CV values**.   From that you could determine the scaling necessary for the second loop to set the blend[] values for the ROI of the image.   It may need a slight adjust if the user recrops but probably that would be the rare case?

da-phil commented:
> @da-phil If you shared the DNG/HDR images where you had the problem I must've missed it. Would you mind sharing them (or telling me how I missed them -- I can be a little slow on github stuff...)? If I had those it would help me understand the problem better.

You're right, I didn't upload the image publicly, I just provided it to @jenshannoschwalm with some other images.

You can generate those HDR DNG files yourself using darktabkes HDR merge feature, by selecting app shots from an exposure bracket series.

I'll see if can share another non-portfolio shot publicly, to serve as a test for this and possibly other features.
Will do once I'm back from the mountains 😉

da-phil commented:
@jenshannoschwalm I just checked the latest version, but get essentially the same result at 35% contrast threshold for the same image

<img width="2408" height="1199" alt="image" src="https://github.com/user-attachments/assets/36c7f214-2a9e-48cf-acf9-d5d9a82955c6" />

I'm just surprised as I understood that you improved handling of very low values in HDR DNG files.

jenshannoschwalm commented:
@da-phil @weltyj @TurboGit after another heavy round of testing on some-100 images with various sensors and exposures there were some minor fixes but i am pretty sure we don't have any signal processing bugs/failures right now. Further improvements (noise reduction or performance gains) would require a) a lot of more work and b) some pixelpipe work - so likely nothing for 5.4.

@TurboGit to be reviewed :-) The second commit is "optional", just let's agree on
1. Do we want a selection menu - i would prefer that as selecting a specific number of iterations is not necessary but i take your word on this.
2. I still prefer **hiding** capture-sharpen-specific parameters if that is disabled as we do that for all other specific options. But again your say.

da-phil commented:
> @jenshannoschwalm @da-phil I think @da-phils confusion stems from the %, i.e. intuitively thinking 35% means 35% of edges are excluded. Maybe the contrast threshold should be renamed, "contrast threshold min < - > max", and show no value to the user.

Not directly that, but the fact that setting the previous threshold default value to 35% gave me considerably less selected pixels than for other "normal" RAW files. But @jenshannoschwalm already explained that this is not (yet) possible due to other work needed in the pipeline.

But there is indeed also another confusion about the meaning of this parameter, as 100% would suggest that 100% of the image content is used for sharpening and lower values restrict sharpening to only high variance image areas.
Maybe we can make this more clear in the parameter description.

jenshannoschwalm commented on src/iop/demosaic.c:
```diff
@@ -340,17 +386,22 @@ int legacy_params(dt_iop_module_t *self,
     } dt_iop_demosaic_params_v2_t;
 
     const dt_iop_demosaic_params_v2_t *o = (dt_iop_demosaic_params_v2_t *)old_params;
-    dt_iop_demosaic_params_v4_t *n = malloc(sizeof(dt_iop_demosaic_params_v4_t));
```
my bad :-) Sure will do

jenshannoschwalm replied:
done

jenshannoschwalm commented:
Force-pushed fixed version
1. corrected module version bump code as requested
2. renamed threshold to "edge sensitivity" as suggested above - also better to me - no "%" any more and improved tooltips
3. hide instead of unsensitive items in first commit.

Leaves out the question: "Use descriptive menu instead of slider for iterations?" 
My personal preference is clear but i take your word on this, as mentioned techy people might love to choose "precisely".

jenshannoschwalm commented:
I see what you mean of course :-) You pinpoint exactly to the weakest point of the algorithm, that's the amplification of chroma/luminance noise and we have a lot of that in this image. You should correct with the edge sensitivity slider excluding those high-noise-low-contrast parts. That will often be around 0.5 with high noise images, also make sure the main radius is not increased too far.

jenshannoschwalm commented:
@weltyj thanks a lot for coworking on this code and your "mask correcting code", indeed more "selectable" and "predictable" especially at sharp luminance transitions. I feared new artefacts but could not spot any and seems to be more stable against chroma noise, very good.

I checked your suggestion of 0.5 as default and prefer it to be a bit lower at 0.4 as that seems more suitable for low noise well exposed images and that seems likely to be the majority.

Fixed as suggested, squashed and force-pushed.

Minor notes: a) using the performance optimized gaussian blur for masks again as we use a sigma of 2.0 which is still perfect for that 9x9 kernel. b) As this code uses the suggested mask mapping we don't check in convolutions kernels for eps but just for above-zero.

weltyj commented:
@jenshannoschwalm Great to hear you see improvements too with your tests.  Thanks for patiently answering my questions too.  You certainly did far, far more on this code so especially thank you for that.   Nice to work with you!

Jeff

weltyj commented:
@da-phil @TurboGit @jenshannoschwalm   I agree with da-phil that the iterations should be a slider.  For two reasons:

1.   I prefer to think of it in terms of iterations because it gives me a numeric sense of what is going on about the power of the sharpening and the computational requirements.
2.  As soon as this goes mainstream, there will be a user that will swear there should be a setting between "average" and "strong".   (or some other adjacent iteration presets)

So, here's a vote to let the user pick the exact iterations with a slider and, if  they want, create their own preset for demosaicing.

jenshannoschwalm commented:
Ok ok, for the sake of swearing users, let's go with the slider :-)

For me it's all good now (made the "sharp centre" visible only if the is a corner boost).

For release notes i would suggest to use the "mini manual" as provided in the commit text

da-phil commented:
As I wrote in [my first post](https://github.com/darktable-org/darktable/pull/18909#issuecomment-3069228678), I don't mind an extra UI element (drop down menu) which enables capture sharpening, if that makes the UI behavior more clear and predictable.
I'd be also fine with @TurboGit's proposal of  keeping the capture sharpening controls inactive and only activate them once iteration count is greater than 0. It's also a way to make the UI behavior more clear and predictable.
So I'm fine with either way, I just want to be able to fine-tune  the iterations with a slider ;)

jenshannoschwalm commented:
i want the UI to be simple, good looking, not distracting from functionality, hiding irrelevant details and not being larger than necessary.

weltyj commented:
Of course, what would be really be great for the UI, is if this looked like a completely separate module, which the user enabled or disabled the normal way.  For that to work, I'm guessing a huge amount of coding magic/rework would be required for demosaic.c to query the state and parameter values of a different module.

The next best thing, I think, is a simple checkbox "enable capture sharpening" in the demosaic.c panel.   When the checkbox is unchecked the capture sharpening elements are hidden.

weltyj commented:
LOVE the UI now!!

Iterations maxing out at 25, vs the previous (20**1.3) ~= 50, does make a visibile difference.  I just fooled around and this

iterations = d->cs_iter + (int)(0.000065f*powf((float)d_cs->iter, 4.0f))


 will give good precision for low iterations (<= 10), and then start ramping up  to get iterations of 50 if d->cs_iter goes to 25

<img width="655" height="565" alt="Screenshot from 2025-07-25 08-17-46" src="https://github.com/user-attachments/assets/95d43606-c935-43c9-84a7-4a68089f601b" />
 ;

weltyj commented:
@jenshannoschwalm I'm not sure what happened but capture.c looks like it didn't change at all after your last update to your github branch...

da-phil commented:
> @TurboGit @da-phil @weltyj friends i think i have sorted it out in an acceptable (for us all) way.
> 
>     1. **All** capture related ui elements have been moved into a collapsible section as used all over the code for such circumstances. Internals parameters are all `gtk_widget_set_sensitive()` if applicable so no fiddling around to-be-visible-or-not.
> 
>     2. Slightly increase resolution of iteration settings - for those of us who think that maths go beyond capabilities of our retina  :-)
> 
>     3. Some improvements in the 'mini manual' and tooltips
> 
> 
> So ready for reviewing pascal.

Thank you so much, this looks perfect now :pray:

jenshannoschwalm commented:
Thanks, there is soon coming more work for you :-)

weltyj commented:
> Thanks, there is soon coming more work for you :-)

Sounds like fun :-)

jenshannoschwalm commented on src/iop/demosaic.c:
```diff
@@ -1244,10 +1378,19 @@ void gui_changed(dt_iop_module_t *self, GtkWidget *w, void *previous)
   gtk_widget_set_visible(g->demosaic_method_bayerfour, bayer4);
   gtk_widget_set_visible(g->demosaic_method_xtrans, xtrans);
 
+  gtk_widget_set_sensitive(g->cs_radius, do_capture);
```
There has been some discussion about this here in this pr. I agree that here in this module we only show applicable parameters ...  The good point with current code that the widget does not grow/shrink so the mouse can always stay where it is is expected to control a slider.

dterrahe replied:
Sorry, didn't see that discussion. How is the situation here _any_ different from all the other places where we hide widgets when they are not applicable? Yes, the gui moves when sliders get hidden. That's why _earlier_ widgets (modes etc) should disable _later_ widgets, not the other way around. Why would this be treated as a completely _new_ issue here, rather than just consistently applying common practice?

jenshannoschwalm replied:
We might want to discuss this in general as a RFC and then decide how to proceed here and filmic.

jenshannoschwalm commented:
Yes, different sensors, different lens, different settings :-)

europlatus commented:
In my testing of this, I have been missing an option to turn it off and on easily. Obviously the Demosaic module can't be turned off, so you just have to slide the Sharpen iterations to zero and then back up again. It's a bit cumbersome. I imagine this was done to reduce clutter, but would it make sense to have an "eye" icon like we have for masks so we can quickly turn the effect off and on again?

jenshannoschwalm commented:
> but would it make sense to have an "eye" icon like we have for masks so we can quickly turn the effect off and on again?

Feel free to implement this :-)

weltyj commented:
but would it make sense to have an "eye" icon like we have for masks so we can quickly turn the effect off and on again

@europlatus A  checkbox "[ ] enable capture sharpening" would be more appropriate than an eye icon.

jenshannoschwalm commented:
Yes, a checkbox might be used/added but i'd prefer to keep the UI as small as possible. If someone wants to propose an idea, feel free to open a PR so we can test/discuss.
